docs: Add Cloudfront invalidation example for S3 deployment (#12238)
aws/aws-cdk: packages/@aws-cdk/aws-codepipeline-actions/README.md

});
```

#### Invalidating the CloudFront cache when deploying to S3

There is currently no native support in CodePipeline for invalidating a CloudFront cache after deployment.
One workaround is to add another build step after the deploy step,
and use the AWS CLI to invalidate the cache:

> ```ts

Reviewer: Maybe add necessary additional import of `import * as iam from '@aws-cdk/aws-iam'`

// Create a Cloudfront Web Distribution
const distribution = new cloudfront.Distribution(this, `Distribution`, {
  // ...
});

// Create the build project that will invalidate the cache
const invalidateBuildProject = new codebuild.PipelineProject(this, `InvalidateProject`, {
  buildSpec: codebuild.BuildSpec.fromObject({
    version: '0.2',
    phases: {
      build: {
        commands:[
          'aws cloudfront create-invalidation --distribution-id ${CLOUDFRONT_ID} --paths "/*"',
          // Choose whatever files or paths you'd like, or all files as specified here
        ],
      },
    },
  }),
  environmentVariables: {
    CLOUDFRONT_ID: { value: distribution.distributionId },
  },
});

// Add Cloudfront invalidation permissions to the project

> const distributionArn = `arn:aws:cloudfront::${this.account}:distribution/${distribution.distributionId}`;

Reviewer: Maybe use `process.env.CDK_DEFAULT_ACCOUNT` instead of `this.account`. The env var for the account should be set automatically as far as I know.

});
```

#### Invalidating the CloudFront cache when deploying to S3

> There is currently no native support in CodePipeline for invalidating a CloudFront cache after deployment.
> 
> One workaround is to add another build step after the deploy step,

Reviewer 2: ```suggestion
There is currently no native support in CodePipeline for invalidating a CloudFront cache after deployment.
One workaround is to add another build step after the deploy step,
```

and use the AWS CLI to invalidate the cache:

```ts
// Create a Cloudfront Web Distribution
const distribution = new cloudfront.Distribution(this, `Distribution`, {
  // ...
});

// Create the build project that will invalidate the cache
const invalidateBuildProject = new codebuild.PipelineProject(this, `InvalidateProject`, {
  buildSpec: codebuild.BuildSpec.fromObject({
    version: '0.2',
    phases: {
      build: {
        commands:[
          'aws cloudfront create-invalidation --distribution-id ${CLOUDFRONT_ID} --paths "/*"',
          // Choose whatever files or paths you'd like, or all files as specified here
        ],
      },
    },
  }),
  environmentVariables: {
    CLOUDFRONT_ID: { value: distribution.distributionId },
  },
});

// Add Cloudfront invalidation permissions to the project
const distributionArn = `arn:aws:cloudfront::${this.account}:distribution/${distribution.distributionId}`;

> invalidateBuildProject.addToRolePolicy(
>   new iam.PolicyStatement({
>     effect: iam.Effect.ALLOW,
>     resources: [distributionArn],
>     actions: [
>       'cloudfront:CreateInvalidation',
>     ],
>   })
> );

Reviewer 2: ```suggestion
invalidateBuildProject.addToRolePolicy(new iam.PolicyStatement({
  resources: [distributionArn],
  actions: [
    'cloudfront:CreateInvalidation',
  ],
}));
```

// Create the pipeline (here only the S3 deploy and Invalidate cache build)
new codepipeline.Pipeline(this, 'Pipeline', {
  stages: [
    // ...
    {
      stageName: 'Deploy',
      actions: [
        new codepipelineActions.S3DeployAction({
          actionName: 'S3Deploy',
          bucket: deployBucket,
          input: deployInput,
          runOrder: 1,
        }),
        new codepipelineActions.CodeBuildAction({
          actionName: 'InvalidateCache',
          project: invalidateBuildProject,

>           input: invalidateOutput,

Reviewer 2: ```suggestion
          input: deployInput,
```

});
```

> #### Invalidating Cloudfront Cache when deploying to S3

Reviewer 2: ```suggestion
#### Invalidating the CloudFront cache when deploying to S3
```

> There is currently no support in the S3DeployAction construct to invalidate the Cloudfront cache after upload.

Reviewer 2: Let's clarify here, as it's not a CDK problem per-se 🙂.

```suggestion
There is currently no native support in CodePipeline for invalidating a CloudFront cache after deployment.
```

> 

Reviewer 2: Let's remove this empty line.

> One workaround is to add another build step after the deploy step, to use the AWS CLI to invalidate the cache.

Reviewer 2: ```suggestion
One workaround is to add another build step after the deploy step,
and use the AWS CLI to invalidate the cache:
```

```ts
// Create a Cloudfront Web Distribution

> const cfDist = new cloudfront.CloudFrontWebDistribution(this, `CloudFrontDistribution`, {

Reviewer 2: `CloudFrontWebDistribution` is the "old" class for this. Let's use the new one, `Distribution`:

```suggestion
const distribution = new cloudfront.Distribution(this, `Distribution`, {
```

>   originConfigs: [
>     {
>       // ...
>       behaviors : [ {isDefaultBehavior: true}]
>     },
>     // ...
>   ],
> });

Reviewer 2: Let's just leave all of this empty - this doc is about CodePipeline, not really about CloudFront 🙂.

```suggestion
 // ...
});
```

// Create the build project that will invalidate the cache
const invalidateBuildProject = new codebuild.PipelineProject(this, `InvalidateProject`, {
  buildSpec: codebuild.BuildSpec.fromObject({
    version: '0.2',
    phases: {
      build: {
        commands:[
          'aws cloudfront create-invalidation --distribution-id ${CLOUDFRONT_ID} --paths "/*"',
          // Choose whatever files or paths you'd like, or all files as specified here
        ],
      },
    },
  }),
  environmentVariables: {
    CLOUDFRONT_ID: { value: cfDist.distributionId },
  },
});

// Add Cloudfront invalidation permissions to the project
invalidateBuildProject.addToRolePolicy(
  new iam.PolicyStatement({
    effect: iam.Effect.ALLOW,

>     resources: ['*'],

Reviewer 2: Hmm, is there a way to reduce these permissions? Do they all have to have `"*"`? Can at least some of them be scoped downed to the Distribution that we're invalidating?

Author: Yup!
I had some old info, that CF didn't have resource level permissions, but they do. So I'm changing so that we only need CreateInvalidation and only for the specific distribution.

    actions: [
      'cloudfront:CreateInvalidation',
      'cloudfront:GetDistribution*',
      'cloudfront:GetInvalidation',
      'cloudfront:ListInvalidations',
      'cloudfront:ListDistributions',
    ],
  })
);

// Create the pipeline (here only the S3 deploy and Invalidate cache build)
new codepipeline.Pipeline(this, 'Pipeline', {
  stages: [
    // ...
    {
      stageName: 'Deploy',
      actions: [
        new codepipelineActions.S3DeployAction({
          actionName: 'S3Deploy',
          bucket: deployBucket,
          input: deployInput,
        }),
        new codepipelineActions.CodeBuildAction({
          actionName: 'InvalidateCache',
          project: invalidateBuildProject,

>           input: invalidateOutput,

Reviewer 2: I think we're missing setting `runOrder` here, no? I think the invalidation needs to happen _after_ the deployment, while as the example is written right now, they will happen at the same time...?

Author: I wasn't aware of that property, now I'm educated about it and will commit a change!
